In [1]:
!pip install transformers torch scikit-learn pydantic matplotlib seaborn nlpaug -q

import pandas as pd
import numpy as np
import random
import torch
from torch.utils.data import Dataset
from transformers import AutoTokenizer, AutoModelForSequenceClassification, Trainer, TrainingArguments
from sklearn.model_selection import train_test_split
from sklearn.metrics import f1_score, precision_score, recall_score, confusion_matrix
import matplotlib.pyplot as plt
import seaborn as sns
from pydantic import BaseModel, Field, ValidationError
import warnings
warnings.filterwarnings('ignore')

print("Окружение готово. GPU:", "Доступен" if torch.cuda.is_available() else "ОШИБКА: Включите GPU!")

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 410.5/410.5 kB 10.9 MB/s eta 0:00:00
Окружение готово. GPU: Доступен


In [3]:
# Наш железобетонный пайплайн из 4-6 уроков
class NomenclatureItem(BaseModel):
    raw_text: str = Field(..., min_length=5)
    target_category: str = Field(..., min_length=2)

file_path = "Выгрузка НП-ЦН 100К.csv"
df_raw = pd.read_csv(file_path, sep=';', encoding='cp1251', on_bad_lines='skip')
df_raw.columns = ['raw_text', 'target_category']
df_raw = df_raw.dropna()

valid_data = []
for idx, row in df_raw.iterrows():
    if "ВЫБЕРИТЕ КОРРЕКТНУЮ ЦН" in str(row['target_category']):
        continue
    try:
        item = NomenclatureItem(
            raw_text=str(row['raw_text']).strip().lower(),
            target_category=str(row['target_category']).strip()
        )
        valid_data.append(item.model_dump())
    except ValidationError:
        pass

df_clean = pd.DataFrame(valid_data)
top_classes = df_clean['target_category'].value_counts().nlargest(50).index
df_baseline = df_clean[df_clean['target_category'].isin(top_classes)].copy()

# Кодируем целевые классы в числа для нейросети
labels_map = {label: i for i, label in enumerate(top_classes)}
reverse_labels_map = {i: label for label, i in labels_map.items()}
df_baseline['label_id'] = df_baseline['target_category'].map(labels_map)

X_train, X_test, y_train, y_test = train_test_split(
    df_baseline['raw_text'].values, df_baseline['label_id'].values,
    test_size=0.2, stratify=df_baseline['label_id'].values, random_state=42
)
print(f"Подготовлено для обучения: {len(X_train)} строк")

Подготовлено для обучения: 8238 строк


In [6]:
import random
import nlpaug.augmenter.char as nac

# Используем RandomCharAug для перестановки соседних букв (action="swap")
aug = nac.RandomCharAug(action="swap", aug_char_p=0.1, aug_char_max=2)

def smart_augment(text, prob=0.3):
    """
    Хирургическая аугментация: вносит опечатки (перестановки) только с вероятностью prob.
    """
    if len(text) < 10 or random.random() > prob:
        return text

    augmented = aug.augment(text)
    return augmented[0] if isinstance(augmented, list) else augmented

# Аугментируем только Train выборку!
X_train_aug = [smart_augment(text, prob=0.3) for text in X_train]

# Для проверки принудительно генерируем 100% опечатку для первой строки:
demo_aug = aug.augment(X_train[0])
demo_aug_text = demo_aug[0] if isinstance(demo_aug, list) else demo_aug

print("Пример умной аугментации (nlpaug):")
print("Оригинал:", X_train[0])
print("С опечаткой:", demo_aug_text)

Пример умной аугментации (nlpaug):
Оригинал: арматура ф 10 а240с гост 34028
С опечаткой: арматуар ф 10 а240с гост 43028


In [7]:
model_name = "cointegrated/rubert-tiny2"
tokenizer = AutoTokenizer.from_pretrained(model_name)

# Создаем Dataset для PyTorch
class NSIDataset(Dataset):
    def __init__(self, texts, labels, tokenizer, max_len=64):
        self.texts = texts
        self.labels = labels
        self.tokenizer = tokenizer
        self.max_len = max_len

    def __len__(self):
        return len(self.texts)

    def __getitem__(self, item):
        text = str(self.texts[item])
        inputs = self.tokenizer(text, max_length=self.max_len, padding='max_length', truncation=True, return_tensors="pt")
        return {
            'input_ids': inputs['input_ids'].flatten(),
            'attention_mask': inputs['attention_mask'].flatten(),
            'labels': torch.tensor(self.labels[item], dtype=torch.long)
        }

train_dataset = NSIDataset(X_train_aug, y_train, tokenizer) # Используем аугментированные!
eval_dataset = NSIDataset(X_test, y_test, tokenizer)

# Инициализируем модель С ГОЛОВОЙ КЛАССИФИКАТОРА (50 классов)
model = AutoModelForSequenceClassification.from_pretrained(model_name, num_labels=50)

# Настройки обучения (HuggingFace Trainer)
training_args = TrainingArguments(
    output_dir='./results',
    num_train_epochs=5,             # 5 эпох для надежности
    per_device_train_batch_size=64, # Оптимально для tiny-версии
    per_device_eval_batch_size=128,
    eval_strategy="epoch",
    save_strategy="epoch",
    learning_rate=3e-5,             # Стандартный шаг для BERT
    weight_decay=0.01,
    load_best_model_at_end=True,
    report_to="none"
)

def compute_metrics(eval_pred):
    logits, labels = eval_pred
    predictions = np.argmax(logits, axis=-1)
    return {'f1_weighted': f1_score(labels, predictions, average='weighted')}

trainer = Trainer(
    model=model,
    args=training_args,
    train_dataset=train_dataset,
    eval_dataset=eval_dataset,
    compute_metrics=compute_metrics
)

print("Начинаю дообучение RuBERT-tiny2... Это займет несколько минут.")
trainer.train()

config.json:   0%|          | 0.00/693 [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/401 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/1.08M [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.74M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/112 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/118M [00:00<?, ?B/s]

Loading weights:   0%|          | 0/55 [00:00<?, ?it/s]

[transformers] BertForSequenceClassification LOAD REPORT from: cointegrated/rubert-tiny2
Key                                        | Status     | 
-------------------------------------------+------------+-
cls.predictions.transform.dense.bias       | UNEXPECTED | 
cls.predictions.transform.LayerNorm.weight | UNEXPECTED | 
cls.seq_relationship.weight                | UNEXPECTED | 
cls.predictions.transform.LayerNorm.bias   | UNEXPECTED | 
cls.predictions.bias                       | UNEXPECTED | 
cls.seq_relationship.bias                  | UNEXPECTED | 
cls.predictions.transform.dense.weight     | UNEXPECTED | 
classifier.weight                          | MISSING    | 
classifier.bias                            | MISSING    | 

Notes:
- UNEXPECTED:	can be ignored when loading from different task/architecture; not ok if you expect identical arch.
- MISSING:	those params were newly initialized because missing from the checkpoint. Consider training on your downstream task.


Начинаю дообучение RuBERT-tiny2... Это займет несколько минут.


Epoch,Training Loss,Validation Loss,F1 Weighted
1,No log,2.232212,0.401205
2,No log,1.890199,0.507548
3,No log,1.718255,0.523096
4,2.136800,1.625538,0.533651
5,2.136800,1.596359,0.557049


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

TrainOutput(global_step=645, training_loss=2.023554484049479, metrics={'train_runtime': 61.2482, 'train_samples_per_second': 672.51, 'train_steps_per_second': 10.531, 'total_flos': 38205582174720.0, 'train_loss': 2.023554484049479, 'epoch': 5.0})

In [8]:
print("--- Оценка модели и Динамическая Постобработка ---")
predictions = trainer.predict(eval_dataset)
logits = predictions.predictions
probs = torch.nn.functional.softmax(torch.tensor(logits), dim=-1).numpy()

# Получаем классы и максимальные вероятности (Confidence)
pred_classes = np.argmax(probs, axis=1)
confidence_scores = np.max(probs, axis=1)

# БИЗНЕС-ЛОГИКА: Нам нужно автоматизировать минимум 85% строк
target_auto_rate = 0.85
total_samples = len(y_test)
target_auto_count = int(total_samples * target_auto_rate)

# Динамически ищем порог, который пропустит ровно 85% самых уверенных предсказаний
sorted_conf = np.sort(confidence_scores)[::-1]
dynamic_threshold = sorted_conf[target_auto_count - 1]

print(f"Бизнес-цель: Автоматизировать {target_auto_rate*100}% позиций.")
print(f"Вычисленный оптимальный порог (Threshold): {dynamic_threshold:.4f}")

final_predictions = []
manual_review_count = 0

for i in range(len(pred_classes)):
    if confidence_scores[i] >= dynamic_threshold:
        final_predictions.append(pred_classes[i])
    else:
        final_predictions.append(-1) # Маркер ручного разбора
        manual_review_count += 1

print(f"Позиций отправлено на ручной разбор экспертом: {manual_review_count} из {total_samples}")

# Считаем метрики только для автоматизированных позиций
mask = np.array(final_predictions) != -1
if np.sum(mask) > 0:
    auto_f1 = f1_score(y_test[mask], np.array(final_predictions)[mask], average='weighted')
    print(f"F1-Weighted на автоматически сопоставленных данных (85% объема): {auto_f1:.4f}")

print("\nВЫВОД ДЛЯ БИЗНЕСА:")
print(f"Мы выполнили KPI: {target_auto_rate*100}% строк обрабатываются автоматически с качеством F1 = {auto_f1:.4f}.")
print("Остаток отправляется на ручной разбор для гарантии отсутствия критических ошибок.")

--- Оценка модели и Динамическая Постобработка ---


Бизнес-цель: Автоматизировать 85.0% позиций.
Вычисленный оптимальный порог (Threshold): 0.0962
Позиций отправлено на ручной разбор экспертом: 309 из 2060
F1-Weighted на автоматически сопоставленных данных (85% объема): 0.6476

ВЫВОД ДЛЯ БИЗНЕСА:
Мы выполнили KPI: 85.0% строк обрабатываются автоматически с качеством F1 = 0.6476.
Остаток отправляется на ручной разбор для гарантии отсутствия критических ошибок.


In [10]:
import torch
import numpy as np
from transformers import AutoTokenizer, AutoModel
from sklearn.svm import LinearSVC
from sklearn.calibration import CalibratedClassifierCV
from sklearn.preprocessing import normalize
from sklearn.metrics import f1_score

print("--- АКТИВАЦИЯ АРХИТЕКТУРЫ ИЗ ПИТЧА ---")
print("1. Эмбеддинги RuBERT-tiny2\n2. L2-нормализация (проекция на гиперсферу)\n3. LinearSVC с балансировкой классов")

# 1. Загружаем чистую базовую модель (без головы классификатора)
embed_tokenizer = AutoTokenizer.from_pretrained("cointegrated/rubert-tiny2")
embed_model = AutoModel.from_pretrained("cointegrated/rubert-tiny2")
embed_model.eval()

if torch.cuda.is_available():
    embed_model = embed_model.to('cuda')

def get_l2_embeddings(texts):
    all_embs = []
    batch_size = 256

    # Принудительно конвертируем входные данные в список строк для защиты от ValueError
    texts_list = list(texts)

    with torch.no_grad():
        for i in range(0, len(texts_list), batch_size):
            batch = texts_list[i:i+batch_size]
            inputs = embed_tokenizer(batch, padding=True, truncation=True, max_length=64, return_tensors="pt")
            if torch.cuda.is_available():
                inputs = {k: v.to('cuda') for k, v in inputs.items()}
            outputs = embed_model(**inputs)

            # Берем [CLS] токен для каждого текста
            cls_embs = outputs.last_hidden_state[:, 0, :].cpu().numpy()
            all_embs.append(cls_embs)

    embs_matrix = np.vstack(all_embs)

    # ТА САМАЯ математическая проекция на L2-гиперсферу из 17-го слайда
    return normalize(embs_matrix, norm='l2')

print("\nГенерация векторов для Train (с опечатками nlpaug) и Test...")
# Массивы X_train_aug и X_test теперь безопасно обрабатываются
X_train_emb = get_l2_embeddings(X_train_aug)
X_test_emb = get_l2_embeddings(X_test.tolist()) # <-- БРОНЕБОЙНОЕ ИСПРАВЛЕНИЕ ЗДЕСЬ

print("\nОбучение LinearSVC (class_weight='balanced')...")
# Используем CalibratedClassifierCV, чтобы SVM научился выдавать вероятности (Confidence Score)
base_svm = LinearSVC(class_weight='balanced', random_state=42, max_iter=2000)
calibrated_svm = CalibratedClassifierCV(base_svm, cv=3)
calibrated_svm.fit(X_train_emb, y_train)

# ОЦЕНКА И ПОСТОБРАБОТКА ДЛЯ БИЗНЕСА
probs = calibrated_svm.predict_proba(X_test_emb)
pred_classes = np.argmax(probs, axis=1)
confidence_scores = np.max(probs, axis=1)

target_auto_rate = 0.85
target_auto_count = int(len(y_test) * target_auto_rate)
dynamic_threshold = np.sort(confidence_scores)[::-1][target_auto_count - 1]

final_predictions = []
for i in range(len(pred_classes)):
    if confidence_scores[i] >= dynamic_threshold:
        final_predictions.append(pred_classes[i])
    else:
        final_predictions.append(-1)

mask = np.array(final_predictions) != -1
auto_f1 = f1_score(y_test[mask], np.array(final_predictions)[mask], average='weighted')

print(f"\n--- ФИНАЛЬНЫЕ МЕТРИКИ ---")
print(f"Вычисленный порог (Threshold) для 85% автоматизации: {dynamic_threshold:.4f}")
print(f"F1-Weighted на автоматически сопоставленных данных (85% объема): {auto_f1:.4f}")

--- АКТИВАЦИЯ АРХИТЕКТУРЫ ИЗ ПИТЧА ---
1. Эмбеддинги RuBERT-tiny2
2. L2-нормализация (проекция на гиперсферу)
3. LinearSVC с балансировкой классов


Loading weights:   0%|          | 0/55 [00:00<?, ?it/s]

[transformers] BertModel LOAD REPORT from: cointegrated/rubert-tiny2
Key                                        | Status     |  | 
-------------------------------------------+------------+--+-
cls.predictions.transform.dense.bias       | UNEXPECTED |  | 
cls.predictions.transform.LayerNorm.weight | UNEXPECTED |  | 
cls.seq_relationship.weight                | UNEXPECTED |  | 
cls.predictions.transform.LayerNorm.bias   | UNEXPECTED |  | 
cls.predictions.bias                       | UNEXPECTED |  | 
cls.seq_relationship.bias                  | UNEXPECTED |  | 
cls.predictions.transform.dense.weight     | UNEXPECTED |  | 

Notes:
- UNEXPECTED:	can be ignored when loading from different task/architecture; not ok if you expect identical arch.



Генерация векторов для Train (с опечатками nlpaug) и Test...

Обучение LinearSVC (class_weight='balanced')...

--- ФИНАЛЬНЫЕ МЕТРИКИ ---
Вычисленный порог (Threshold) для 85% автоматизации: 0.5505
F1-Weighted на автоматически сопоставленных данных (85% объема): 0.8945


### Отчет по оптимизации пайплайна и бизнес-метрик

В ходе доработки проекта был проведен масштабный рефакторинг пайплайна с фокусом на жесткие требования бизнеса и отказоустойчивость модели:

1. **Хирургическая аугментация (`nlpaug`):** Классические методы обработки естественного языка (случайное удаление или замена слов) деструктивны для технической номенклатуры, где потеря одной цифры меняет марку стали или ГОСТ. Внедрена библиотека `nlpaug` (метод `RandomCharAug` с параметром `action="swap"`). Данный метод реалистично имитирует слепые опечатки оператора (перестановку соседних символов при быстром наборе), сохраняя при этом общую структуру артикула для обучения.

2. **Конфликт бизнес-метрик и классического ML:** Базовая бизнес-цель системы — автоматизировать **85%** рутинной обработки входящих строк. Эксперимент показал, что при использовании стандартной "головы" классификатора над трансформером, для захвата 85% потока алгоритму требуется опустить Confidence Threshold до экстремальных **0.0962**. При таком низком пороге уверенности метрика $F_1\text{-weighted}$ предсказуемо падает до **0.6476**, что несет недопустимые риски для качества корпоративных справочников.

3. **Переход на гибридную архитектуру (Математическое ядро):** Для решения проблемы сильного дисбаланса классов и удержания метрик, пайплайн был переведен на более устойчивую гибридную архитектуру:
    * Генерация сырых эмбеддингов через базовую модель `RuBERT-tiny2`.
    * Применение $L_2$-нормализации векторов эмбеддингов для устранения искажений дистанции в косинусном пространстве:
    
    $$ \mathbf{v}_{norm} = \frac{\mathbf{v}}{|\mathbf{v}|2} = \frac{\mathbf{v}}{\sqrt{\sum{i=1}^{n} v_i^2}} $$
    
    Данная проекция признаков на единичную гиперсферу необходима для корректной работы алгоритмов, максимизирующих разделяющую гиперплоскость.
    * Классификация через `LinearSVC` с параметром `class_weight='balanced'`, жестко штрафующим модель за ошибки на миноритарных классах.

4. **Итоговый продуктовый результат:** Внедренный динамический алгоритм калибровки установил оптимальный порог автоматизации на уровне **0.5505**. При данном пороге уверенности в автоматический разбор уходит заданный бизнес-KPI в **85%** строк, а итоговое качество разметки составило **$F_1\text{-weighted} = 0.8945$**. Оставшиеся 15% (наиболее сложные, пограничные или принципиально новые товарные позиции) корректно блокируются системой и маршрутизируются на ручной разбор специалисту.